authenticate to LDAP with client certificates

zmb3 opened the pull request:
Rather than requiring a password for the LDAP service account,
we can use a Teleport-issued certificate to authenticate.

This works because AD must already be configured to trust the Teleport
CA in order for desktop access to function.

Fixes #8921

zmb3 committed:
authenticate to LDAP with client certificates

Rather than requiring a password for the LDAP service account,
we can use a Teleport-issued certificate to authenticate.

This works because AD must already be configured to trust the Teleport
CA in order for desktop access to function.

Fixes #8921

zmb3 committed:
Remove mention of LDAP password from docs

Teleport will issue itself a certificate for the configured LDAP
user in order to authenticate.

zmb3 commented:
I'm thinking I will _not_ backport this, and just make this behavior specific to Teleport 9+, while the desktop access preview in Teleport 8.x will continue to require `password_file`. This way we don't have to have special notes in the docs for the specific patch release where we made the change.

Let me know if this doesn't make sense.

zmb3 commented:
Note: this appears to work with Windows Server 2016 and later. We're still investigating if there's something we can do on Windows Server 2012.

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -52,8 +46,8 @@ do {
 
-# Save the plaintext password to a file for later use in your teleport.yaml.
-$Password | Out-File $OutputFile
+Write-Host "Generated password for $SamAccountName: $Password"
```
Good point, maybe we just discard the password entirely. An admin can always reset the password if they want to log in using this account.

zmb3 replied:
Updated, we'll just discard the password now.

smallinsky commented on lib/srv/desktop/windows_server.go:
```diff
@@ -355,6 +366,71 @@ func NewWindowsService(cfg WindowsServiceConfig) (*WindowsService, error) {
+		user = user[i+1:]
+	}
+
+	certDER, keyDER, err := s.generateCredentials(context.Background(), user, s.cfg.Domain, windowsDesktopServiceCertTTL)
```
```suggestion
	certDER, keyDER, err := s.generateCredentials(s.closeCtx, user, s.cfg.Domain, windowsDesktopServiceCertTTL)
```

smallinsky commented on lib/srv/desktop/ldap.go:
```diff
@@ -17,64 +17,43 @@ limitations under the License.
+func (c *ldapClient) setClient(client ldap.Client) {
+	c.mu.Lock()
+	if c.client != nil {
+		c.client.Close()
```
If a client is returned by `ldapClient.c()` and then the setClient will be  called the next use of the client returned by the `ldapClient.c()` will  failed because the client was closed.

smallinsky commented on lib/srv/desktop/windows_server.go:
```diff
@@ -355,6 +366,71 @@ func NewWindowsService(cfg WindowsServiceConfig) (*WindowsService, error) {
+
+	// ensure that we refresh our LDAP cert well before expiration
+	go func() {
+		<-time.After(windowsDesktopServiceCertTTL / 3)
```
```suggestion
               switch {
               case <-s.closecCtx.Done(): 
                  return
		case <-time.After(windowsDesktopServiceCertTTL / 3):
		}
		...
		
```

smallinsky commented on lib/srv/desktop/windows_server.go:
```diff
@@ -272,14 +273,20 @@ func NewWindowsService(cfg WindowsServiceConfig) (*WindowsService, error) {
+	clusterName, err := cfg.AccessPoint.GetClusterName()
 	if err != nil {
-		return nil, trace.Wrap(err, "connecting to LDAP server: %v", err)
+		return nil, trace.Wrap(err, "fetching cluster name: %v", err)
```
```suggestion
		return nil, trace.Wrap(err, "fetching cluster name: %v", clusterName)
```

smallinsky reviewed (commented):
Left some minor comments

I have one question regarding: 
> make this behavior specific to Teleport 9+, while the desktop access preview in Teleport 8.x will continue to require password_file.

Can we just drop old password auth in teleport 9 without marking it as deprecated with simultaneous support with CA based ldap login   ?  Though that we should support +1 teleport major version config compatibility.

zmb3 commented:
> Can we just drop old password auth in teleport 9 without marking it as deprecated with simultaneous support with CA based ldap login ? Though that we should support +1 teleport major version config compatibility.

Yes, that was what I was thinking.

As for config compatibility, I guess we can just leave the field in the config file, and if we detect that it is set log a warning that it won't be used. (this assuming I'm able to get certificate auth working with Windows Server 2012)

zmb3 committed:
Run LDAP initialization in a retry loop

Prior to this change, the entire Teleport process would crash if
it is unable to connect to LDAP on startup.

Instead, we perform retries so that the service doesn't fail if
an LDAP client cannot be initialized. Initialization failures could
occur due to temporary outages of the LDAP server, misconfiguration,
or because Teleport Enterprise is running and the auth server is not
licensed to generate Windows certs.

This has the added side effect of re-publishing the Windows CRL every
time we re-issue the LDAP client certificate, which will ensure it stays
up to date, even in the face of CA rotations.

Fixes #8527

klizhentas commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -255,10 +285,13 @@ gpupdate.exe /force
+let Teleport know that the certificate sent by your Domain Controller during the
+initial SSL connection is trusted. If your Domain Controller's certificate is
+trusted by the system repository on the system running Teleport, you can skip
+this step. If you are unable to acquire the LDAP CA certificate, you can skip
```
Nit: I would hide this sentence in a hidden details block

xinding33 commented on docs/pages/includes/desktop-access/desktop-config.mdx:
```diff
@@ -17,15 +17,13 @@ windows_desktop_service:
+    # likely "EXAMPLE". When connecting as the "svc-teleport" user, you should
+    # use the format: "EXAMPLE\svc-teleport".
     username: '$LDAP_USERNAME'
-    # Plain text file containing the LDAP password for authentication.
-    # This is usually the same password you use to login to the Domain Controller.
-    password_file: /var/lib/ldap-pass
```
I'm a big fan of removing the `password_file` from a security perspective because clearly plaintext passwords are not secure. I might have missed it, but what error are we returning to users who leave behind this file in their config? Given that this will be the major release where we are bringing desktop access out of preview, I'm not worried about phasing out existing users so I'm onboard with erroring out (as opposed to something less aggressive such as a warning).

xinding33 replied:
What do you think about creating a special case for this? Perhaps something more descriptive like "field password_file was deprecated in Teleport 9.0.0 in favor of certificate-based authentication, see {link} for more information. Delete password_file from your teleport.yaml file to fix this error."

We can link to the relevant 9.0.0 release notes where we cover the details.

I think this is an edge case so I'm comfortable with keeping the existing behavior, especially if we are against adding special cases.

zmb3 replied:
I'm fine with that, would be a nicer experience for the user. We can do this and schedule it for deletion in Teleport 10.

xinding33 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -295,7 +329,6 @@ windows_desktop_service:
     addr: "$LDAP_SERVER_ADDRESS"
     domain: "$LDAP_DOMAIN_NAME"
     username: '$LDAP_USERNAME'
-    password_file: /var/lib/teleport-svc-pass.txt
     # This should be the path to the certificate exported in Step 4.
```
Where do we typically document feature deprecation? I would like to mention somewhere in our documentation that `password_file` is being deprecated in favor of a superior implementation.

zmb3 replied:
This seems like a good thing to flag in the release notes.

xinding33 replied:
Agreed, that makes sense.

xinding33 reviewed (commented):
I'm a big fan of removing the need for an LDAP password for desktop access! I'm also in favor of deprecating the `password_file` and erroring out if users leave it behind in their config. I left a few comments with clarifying questions.

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -184,6 +191,28 @@ Computer Configuration > Policies > Windows Settings > Security Settings > Publi
   ![Import Teleport CA](../../img/desktop-access/ca.png)
 </Figure>
 
+### Publish the Teleport CA to the NTAuth Store
```
Yes, Teleport used to do this for you using the password you provide it.

Now that there's no longer a password you need to do this before you start Teleport for the first time.

(From that point on Teleport will issue itself a certificate and renew it about every 6 hours)

zmb3 committed:
Check if the legacy password_file config field is set

This change will help users who upgrade from Teleport 8 to Teleport 9
by giving them a clear and actionable error message rather than the
default "field password_file not found in type config.LDAPConfig".

Marked for removal in Teleport 10.

zmb3 committed:
Put note about skipping TLS verification in a <Details> box
